Use WebProcess::singleton instead of unnecessarily storing a duplicate pointer

achristensen07 opened the pull request:
#### ca61c4e9a1033605f980bd6a33188e8bd5c08e5f
<pre>
Use WebProcess::singleton instead of unnecessarily storing a duplicate pointer
<a href="https://bugs.webkit.org/show_bug.cgi?id=257778">https://bugs.webkit.org/show_bug.cgi?id=257778</a>
rdar://110364959

Reviewed by Michael Catanzaro.

This reduces memory use.  I&apos;m not aware of any reason why we should store duplicate pointers
to a singleton object.  I&apos;m not aware of plans to make it not a singleton.

We need to still pass WebProcess&amp; as a parameter to functions called from the WebProcess constructor,
or we get infinite recursion in the singleton call.

* Source/WebKit/WebProcess/GPU/media/RemoteAudioHardwareListener.cpp:
(WebKit::RemoteAudioHardwareListener::create):
(WebKit::RemoteAudioHardwareListener::RemoteAudioHardwareListener):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioHardwareListener.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::create):
(WebKit::RemoteAudioSession::RemoteAudioSession):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/GPU/media/RemoteCDMFactory.cpp:
(WebKit::RemoteCDMFactory::RemoteCDMFactory):
(WebKit::RemoteCDMFactory::gpuProcessConnection):
* Source/WebKit/WebProcess/GPU/media/RemoteCDMFactory.h:
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVFManager.cpp:
(WebKit::RemoteImageDecoderAVFManager::RemoteImageDecoderAVFManager):
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVFManager.h:
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.cpp:
(WebKit::RemoteLegacyCDMFactory::RemoteLegacyCDMFactory):
(WebKit::RemoteLegacyCDMFactory::gpuProcessConnection):
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaEngineConfigurationFactory.cpp:
(WebKit::RemoteMediaEngineConfigurationFactory::RemoteMediaEngineConfigurationFactory):
(WebKit::RemoteMediaEngineConfigurationFactory::gpuProcessConnection):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaEngineConfigurationFactory.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::RemoteMediaPlayerManager):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.h:
(WebKit::RemoteMediaPlayerManager::parentProcess const): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteRemoteCommandListener.cpp:
(WebKit::RemoteRemoteCommandListener::create):
(WebKit::RemoteRemoteCommandListener::RemoteRemoteCommandListener):
* Source/WebKit/WebProcess/GPU/media/RemoteRemoteCommandListener.h:
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::RemoteMediaSessionHelper):
(WebKit::RemoteMediaSessionHelper::ensureConnection):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:
* Source/WebKit/WebProcess/Geolocation/WebGeolocationManager.cpp:
(WebKit::WebGeolocationManager::WebGeolocationManager):
(WebKit::WebGeolocationManager::registerWebPage):
(WebKit::WebGeolocationManager::unregisterWebPage):
(WebKit::WebGeolocationManager::setEnableHighAccuracyForPage):
* Source/WebKit/WebProcess/Geolocation/WebGeolocationManager.h:
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::WebNotificationManager::WebNotificationManager):
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::transformHandlesToObjects):
(WebKit::WebProcess::setUseGPUProcessForMedia):
* Source/WebKit/WebProcess/cocoa/AudioSessionRoutingArbitrator.cpp:
(WebKit::AudioSessionRoutingArbitrator::AudioSessionRoutingArbitrator):
(WebKit::AudioSessionRoutingArbitrator::beginRoutingArbitrationWithCategory):
(WebKit::AudioSessionRoutingArbitrator::leaveRoutingAbritration):
(WebKit::AudioSessionRoutingArbitrator::canLog const):
* Source/WebKit/WebProcess/cocoa/AudioSessionRoutingArbitrator.h:
* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.cpp:
(WebKit::UserMediaCaptureManager::UserMediaCaptureManager):
(WebKit::UserMediaCaptureManager::~UserMediaCaptureManager):
* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.h:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::transformHandlesToObjects):
* Source/WebKit/WebProcess/glib/UserMediaCaptureManager.cpp:
(WebKit::UserMediaCaptureManager::UserMediaCaptureManager):
(WebKit::UserMediaCaptureManager::~UserMediaCaptureManager):
* Source/WebKit/WebProcess/glib/UserMediaCaptureManager.h:

Canonical link: <a href="https://commits.webkit.org/265274@main">https://commits.webkit.org/265274@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53f8362149d3fcf644940569cffd7d2e233e4034

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10454 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10683 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10958 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12094 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10026 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12938 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10639 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12986 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10613 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11493 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8802 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12493 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8792 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16720 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9870 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9610 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/12860 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10048 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8151 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9206 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2509 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13462 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9909 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/WebProcess/glib/UserMediaCaptureManager.h:
```diff
@@ -46,7 +46,7 @@ class UserMediaCaptureManager : public WebProcessSupplement, public IPC::Message
     WTF_MAKE_NONCOPYABLE(UserMediaCaptureManager);
 public:
-    explicit UserMediaCaptureManager(WebProcess&);
+    explicit UserMediaCaptureManager();
```
```suggestion
    UserMediaCaptureManager();
```

mcatanzaro commented on Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.h:
```diff
@@ -49,7 +49,7 @@ class WebProcess;
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit UserMediaCaptureManager(WebProcess&);
+    explicit UserMediaCaptureManager();
```
```suggestion
    UserMediaCaptureManager();
```

mcatanzaro commented on Source/WebKit/WebProcess/cocoa/AudioSessionRoutingArbitrator.h:
```diff
@@ -40,7 +40,7 @@ class AudioSessionRoutingArbitrator final
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit AudioSessionRoutingArbitrator(WebProcess&);
+    explicit AudioSessionRoutingArbitrator();
```
```suggestion
    AudioSessionRoutingArbitrator();
```

mcatanzaro commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -46,14 +46,14 @@ class WebProcess;
 
 private:
-    explicit WebPermissionController(WebProcess&);
+    explicit WebPermissionController();
```
```suggestion
    WebPermissionController();
```

mcatanzaro commented on Source/WebKit/WebProcess/WebCoreSupport/RemoteWebLockRegistry.h:
```diff
@@ -51,7 +51,7 @@ class RemoteWebLockRegistry final : public WebCore::WebLockRegistry, public IPC:
 
 private:
-    explicit RemoteWebLockRegistry(WebProcess&);
+    explicit RemoteWebLockRegistry();
```
```suggestion
    RemoteWebLockRegistry();
```

mcatanzaro commented on Source/WebKit/WebProcess/Notifications/WebNotificationManager.h:
```diff
@@ -53,7 +53,7 @@ class WebNotificationManager : public WebProcessSupplement, public IPC::MessageR
     WTF_MAKE_NONCOPYABLE(WebNotificationManager);
 public:
-    explicit WebNotificationManager(WebProcess&);
+    explicit WebNotificationManager();
```
```suggestion
    WebNotificationManager();
```

mcatanzaro commented on Source/WebKit/WebProcess/MediaCache/WebMediaKeyStorageManager.h:
```diff
@@ -42,7 +42,7 @@ class WebMediaKeyStorageManager : public WebProcessSupplement {
     WTF_MAKE_NONCOPYABLE(WebMediaKeyStorageManager);
 public:
-    explicit WebMediaKeyStorageManager(WebProcess&) { }
+    explicit WebMediaKeyStorageManager() { }
```
```suggestion
    WebMediaKeyStorageManager() { }
```

mcatanzaro commented on Source/WebKit/WebProcess/Geolocation/WebGeolocationManager.h:
```diff
@@ -48,7 +48,7 @@ class WebGeolocationManager : public WebProcessSupplement, public IPC::MessageRe
     WTF_MAKE_NONCOPYABLE(WebGeolocationManager);
 public:
-    explicit WebGeolocationManager(WebProcess&);
+    explicit WebGeolocationManager();
```
```suggestion
    WebGeolocationManager();
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.h:
```diff
@@ -37,7 +37,7 @@ class WebProcess;
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit RemoteVideoCodecFactory(WebProcess&);
+    explicit RemoteVideoCodecFactory();
```
```suggestion
    RemoteVideoCodecFactory();
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteRemoteCommandListener.h:
```diff
@@ -48,8 +48,8 @@ class RemoteRemoteCommandListener final
-    static Ref<RemoteRemoteCommandListener> create(WebCore::RemoteCommandListenerClient&, WebProcess&);
-    RemoteRemoteCommandListener(WebCore::RemoteCommandListenerClient&, WebProcess&);
+    static Ref<RemoteRemoteCommandListener> create(WebCore::RemoteCommandListenerClient&);
+    RemoteRemoteCommandListener(WebCore::RemoteCommandListenerClient&);
```
This time, the opposite:

```suggestion
    explicit RemoteRemoteCommandListener(WebCore::RemoteCommandListenerClient&);
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.h:
```diff
@@ -51,11 +51,10 @@ class RemoteMediaPlayerManager
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit RemoteMediaPlayerManager(WebProcess&);
+    explicit RemoteMediaPlayerManager();
```
```suggestion
    RemoteMediaPlayerManager();
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteMediaEngineConfigurationFactory.h:
```diff
@@ -49,13 +49,12 @@ class RemoteMediaEngineConfigurationFactory final
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit RemoteMediaEngineConfigurationFactory(WebProcess&);
+    explicit RemoteMediaEngineConfigurationFactory();
```
```suggestion
    RemoteMediaEngineConfigurationFactory();
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.h:
```diff
@@ -61,13 +61,12 @@ class RemoteLegacyCDMFactory final
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit RemoteLegacyCDMFactory(WebProcess&);
+    explicit RemoteLegacyCDMFactory();
```
```suggestion
    RemoteLegacyCDMFactory();
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVFManager.h:
```diff
@@ -49,7 +49,7 @@ class RemoteImageDecoderAVFManager final
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit RemoteImageDecoderAVFManager(WebProcess&);
+    explicit RemoteImageDecoderAVFManager();
```
```suggestion
    RemoteImageDecoderAVFManager();
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteCDMFactory.h:
```diff
@@ -57,11 +57,10 @@ class RemoteCDMFactory final
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    explicit RemoteCDMFactory(WebProcess&);
+    explicit RemoteCDMFactory();
```
```suggestion
    RemoteCDMFactory();
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
```diff
@@ -49,16 +49,16 @@ class RemoteAudioSession final
-    friend UniqueRef<RemoteAudioSession> WTF::makeUniqueRefWithoutFastMallocCheck<RemoteAudioSession>(WebProcess&);
-    explicit RemoteAudioSession(WebProcess&);
+    friend UniqueRef<RemoteAudioSession> WTF::makeUniqueRefWithoutFastMallocCheck<RemoteAudioSession>();
+    explicit RemoteAudioSession();
```
```suggestion
    RemoteAudioSession();
```

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteAudioHardwareListener.h:
```diff
@@ -48,15 +48,15 @@ class RemoteAudioHardwareListener final
 
 private:
-    RemoteAudioHardwareListener(WebCore::AudioHardwareListener::Client&, WebProcess&);
+    RemoteAudioHardwareListener(WebCore::AudioHardwareListener::Client&);
```
```suggestion
    explicit RemoteAudioHardwareListener(WebCore::AudioHardwareListener::Client&);
```

mcatanzaro reviewed (commented):
Well something broke somewhere, although I cannot guess what. It _seems_ like it should have been a safe change.

Anyway, my only comment here is to update use of `explicit` when changing the number of parameters of the constructor. There are two places where it should be added, and a lot where it should be removed. (There's also one place where it was previously desired but missing, which you fixed.)

mcatanzaro commented on Source/WebKit/WebProcess/GPU/media/RemoteCDMFactory.cpp:
```diff
@@ -39,8 +39,7 @@ namespace WebKit {
 
-RemoteCDMFactory::RemoteCDMFactory(WebProcess& process)
-    : m_process(process)
+RemoteCDMFactory::RemoteCDMFactory(WebProcess&)
```
Wait, why did you keep the parameter here? This looks like an oversight.

mcatanzaro replied:
OK, I see you've done this for classes that inherit from WebProcessSupplement, where it's needed. ack.

achristensen07 committed:
Use WebProcess::singleton instead of unnecessarily storing a duplicate pointer
https://bugs.webkit.org/show_bug.cgi?id=257778
rdar://110364959

Reviewed by Michael Catanzaro.

This reduces memory use.  I'm not aware of any reason why we should store duplicate pointers
to a singleton object.  I'm not aware of plans to make it not a singleton.

We need to still pass WebProcess& as a parameter to functions called from the WebProcess constructor,
or we get infinite recursion in the singleton call.

* Source/WebKit/WebProcess/GPU/media/RemoteAudioHardwareListener.cpp:
(WebKit::RemoteAudioHardwareListener::create):
(WebKit::RemoteAudioHardwareListener::RemoteAudioHardwareListener):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioHardwareListener.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::create):
(WebKit::RemoteAudioSession::RemoteAudioSession):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/GPU/media/RemoteCDMFactory.cpp:
(WebKit::RemoteCDMFactory::RemoteCDMFactory):
(WebKit::RemoteCDMFactory::gpuProcessConnection):
* Source/WebKit/WebProcess/GPU/media/RemoteCDMFactory.h:
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVFManager.cpp:
(WebKit::RemoteImageDecoderAVFManager::RemoteImageDecoderAVFManager):
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVFManager.h:
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.cpp:
(WebKit::RemoteLegacyCDMFactory::RemoteLegacyCDMFactory):
(WebKit::RemoteLegacyCDMFactory::gpuProcessConnection):
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaEngineConfigurationFactory.cpp:
(WebKit::RemoteMediaEngineConfigurationFactory::RemoteMediaEngineConfigurationFactory):
(WebKit::RemoteMediaEngineConfigurationFactory::gpuProcessConnection):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaEngineConfigurationFactory.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::RemoteMediaPlayerManager):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.h:
(WebKit::RemoteMediaPlayerManager::parentProcess const): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteRemoteCommandListener.cpp:
(WebKit::RemoteRemoteCommandListener::create):
(WebKit::RemoteRemoteCommandListener::RemoteRemoteCommandListener):
* Source/WebKit/WebProcess/GPU/media/RemoteRemoteCommandListener.h:
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::RemoteMediaSessionHelper):
(WebKit::RemoteMediaSessionHelper::ensureConnection):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:
* Source/WebKit/WebProcess/Geolocation/WebGeolocationManager.cpp:
(WebKit::WebGeolocationManager::WebGeolocationManager):
(WebKit::WebGeolocationManager::registerWebPage):
(WebKit::WebGeolocationManager::unregisterWebPage):
(WebKit::WebGeolocationManager::setEnableHighAccuracyForPage):
* Source/WebKit/WebProcess/Geolocation/WebGeolocationManager.h:
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::WebNotificationManager::WebNotificationManager):
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::transformHandlesToObjects):
(WebKit::WebProcess::setUseGPUProcessForMedia):
* Source/WebKit/WebProcess/cocoa/AudioSessionRoutingArbitrator.cpp:
(WebKit::AudioSessionRoutingArbitrator::AudioSessionRoutingArbitrator):
(WebKit::AudioSessionRoutingArbitrator::beginRoutingArbitrationWithCategory):
(WebKit::AudioSessionRoutingArbitrator::leaveRoutingAbritration):
(WebKit::AudioSessionRoutingArbitrator::canLog const):
* Source/WebKit/WebProcess/cocoa/AudioSessionRoutingArbitrator.h:
* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.cpp:
(WebKit::UserMediaCaptureManager::UserMediaCaptureManager):
(WebKit::UserMediaCaptureManager::~UserMediaCaptureManager):
* Source/WebKit/WebProcess/cocoa/UserMediaCaptureManager.h:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::transformHandlesToObjects):
* Source/WebKit/WebProcess/glib/UserMediaCaptureManager.cpp:
(WebKit::UserMediaCaptureManager::UserMediaCaptureManager):
(WebKit::UserMediaCaptureManager::~UserMediaCaptureManager):
* Source/WebKit/WebProcess/glib/UserMediaCaptureManager.h:

Canonical link: https://commits.webkit.org/265274@main
